### PR TITLE
Switch to a system with 3 modes for values

### PR DESCRIPTION
**Pull request by lpw25**
This switches to the 3 mode system that we've been discussing. This allows us distinguish values from the current region frame from values in parent regions. This allows us to apply the restrictions on function returns and tailcalls only to values from the current region.

As an example, this allows:
```ocaml
let rec length l =
  let rec loop acc = function
    | [] -> 0
    | x :: xs -> loop (acc + 1) xs
  in
  loop 0 l
```
to be given the type:
```ocaml
val length : local_ 'a list -> int
```

### CODE REVIEW

File: typing/parmatch.ml [hunk 1]
```diff
@@ -38,7 +38,7 @@ let omega_list = Patterns.omega_list
 let extra_pat =
   make_pat
     (Tpat_var (Ident.create_local "+", mknoloc "+"))
-    Ctype.none Alloc_mode.local Env.empty
+    Ctype.none Value_mode.min_mode Env.empty
```
Comment (stedolan) [inline typing/parmatch.ml]: Is this intentional? `Value_mode.min_mode` is global, not local. (Same question for the other changes in this file and patterns.ml)

File: typing/types.mli [hunk 2]
```diff
@@ -600,29 +601,154 @@ module Alloc_mode : sig
 
   (* Modes are ordered so that [heap] is a submode of [local] *)
```
Comment (stedolan) [inline typing/types.mli]: `[heap]` not renamed here

File: typing/types.ml [hunk 3]
```diff
@@ -589,6 +590,11 @@ module Alloc_mode = struct
     | () -> Ok ()
     | exception NotSubmode -> Error ()
 
+  let submode_exn t1 t2 =
+    match submode t1 t2 with
+    | Ok () -> ()
+    | Error () -> assert false
```
Comment (stedolan) [inline typing/types.ml]: `assert false` isn't what I expect from the failure case of a function called `foo_exn`

File: typing/types.ml [hunk 4]
```diff
@@ -648,4 +669,163 @@ module Alloc_mode = struct
...
+        ~(nonlocal : Alloc_mode.const) ~(nonregional : Alloc_mode.const) =
+    match nonlocal, nonregional with
+    | Global, Global -> Global
+    | Global, Local -> assert false
+    | Local, Global -> Regional
+    | Local, Local -> Local
+
+  type t =
+    { nonlocal : Alloc_mode.t;
+      nonregional : Alloc_mode.t; }
```
Comment (stedolan) [inline typing/types.ml]: A comment about the invariant that `nonlocal >= nonregional` would help.

File: typing/types.ml [hunk 5]
```diff
@@ -620,26 +639,28 @@ module Alloc_mode = struct
...
+      end
+    and trans_low v' =
+      submode_cv v'.lower v;
+      List.iter trans v'.vlower
+    in
+    List.iter trans_low v.vlower
+
   let constrain_lower = function
     | Amode m -> m
-    | Amodevar v when v.lower = v.upper -> v.lower
```
Comment (stedolan) [inline typing/types.ml:625]: Why was this removed?

File: typing/env.ml [hunk 6]
```diff
@@ -2824,9 +2849,7 @@ let lookup_all_labels_from_type ?(use=true) ~loc ty_path env =
 
 let lookup_instance_variable ?(use=true) ~loc name env =
   match IdTbl.find_name_and_modes wrap_value ~mark:use name env.values with
-  | (path, locks, Val_bound vda) -> begin
-      constrain_modes ~errors:true ~loc env (Lident name)
-        vda.vda_mode locks Alloc_mode.heap;
+  | (path, _, Val_bound vda) -> begin
```
Comment (stedolan) [inline typing/env.ml:2842]: Why is it OK to ignore the locks here? Is it because instance variables are always global?

File: typing/env.ml [hunk 7]
```diff
@@ -2414,21 +2425,31 @@ let lookup_ident_module (type a) (load : a load) ~errors ~use ~loc s env =
...
+      | Region_lock -> ()
+      | Lock { mode } -> begin
+          match Alloc_mode.submode vmode_nonlocal mode with
+          | Ok () -> ()
+          | Error () ->
+            may_lookup_error errors loc env
+              (Local_value_used_in_closure id)
+        end)
+    locks;
+  vmode
```
Comment (stedolan) [inline typing/env.ml]: This code reorders locks by processing all of the region locks first. This is suspicious, but works here because region locks only affect the distinction between `Regional` and `Local`, which is ignored by other locks.

This is quite subtle. To make it more obviously correct, would you consider processing the locks in order as a fold?

File: typing/env.mli [hunk 8]
```diff
@@ -186,8 +186,8 @@ val lookup_error: Location.t -> t -> lookup_error -> 'a
    emitted the wrong number of times. *)
 
 val lookup_value:
-  ?use:bool -> loc:Location.t -> Longident.t -> Types.alloc_mode -> t ->
-  Path.t * value_description
+  ?use:bool -> loc:Location.t -> Longident.t -> t ->
+  Path.t * value_description * Types.Value_mode.t
```
Comment (stedolan) [inline typing/env.mli:190]: This is a definite improvement to the API.

File: typing/typecore.ml [hunk 9]
```diff
@@ -4543,10 +4756,12 @@ and type_application env app_loc (mode : expected_mode) funct sargs =
               raise(Error(funct.exp_loc, env, Apply_non_function
                             (expand_head env funct.exp_type)))
     in
+    let m_arg = Alloc_mode.newvar () in
+    Alloc_mode.submode_exn m_arg m_param;
```
Comment (stedolan) [inline typing/typecore.ml:4765]: What's the fresh variable for? (Or, why is it better to have `m_arg <= m_param` instead of `m_arg = m_param`?)

File: typing/types.mli [hunk 10]
```diff
@@ -600,29 +601,154 @@ module Alloc_mode : sig
...
+
+  (** Right adjoint of [of_alloc_nonregional] and left adjoint of
+      [of_alloc_nonglobal].
+
+     {[
+       Global --> Global
+       Regional --> Global
+       Local --> Local
+     ]} *)
+  val to_alloc_nonregional : t -> Alloc_mode.t
```
Comment (stedolan) [inline typing/types.mli]: The names of the two `to_alloc` functions are extremely confusing, I had to keep referring to their declarations. (I'm not sure what better ones would be, sadly!)

File: testsuite/tests/typing-local/local.ml [hunk 11]
```diff
@@ -170,47 +162,55 @@ val apply4 : int -> int = <fun>
...
 [%%expect{|
-val app43 :
-  (a:local_ int ref -> (int -> b:local_ int ref -> c:int -> unit)) ->
-  b:local_ int ref -> c:int -> unit = <fun>
+Line 2, characters 7-21:
+2 |   f ~a:(local_ ref 1) 2
+           ^^^^^^^^^^^^^^
+Error: This local value escapes its region
+  Hint: Local values cannot be arguments to tail-calls
+  unless the result of the tail-call has a "local_" annotation
```
Comment (stedolan) [inline testsuite/tests/typing-local/local.ml]: This change is slightly surprising to me. (I think both behaviours are fine)

File: typing/typecore.ml [hunk 12]
```diff
@@ -5532,6 +5768,25 @@ let report_type_expected_explanation expl ppf =
...
+let escaping_hint reason context =
+  match reason, context with
+  | Types.Value_mode.Locality, Some Return ->
+      [ Location.msg
+          "@[Hint: Cannot return local value without an explicit@ \
+           \"local_\" annotation@]" ]
+  | Types.Value_mode.Locality, Some Tailcall_argument ->
+      [ Location.msg
+          "@[Hint: Local values cannot be arguments to tail-calls@ \
+           unless the result of the tail-call has a \"local_\" annotation@]" ]
```
Comment (stedolan) [inline typing/typecore.ml]: I think we should reword this hint. From some experimentation, this is an extremely common issue. This hint, while accurate, suggests that the fix is to add a `local_` annotation to the result of the called function. This is almost never the right thing to do: usually, you should make the call non-tail instead.

Suggestion:
```
This argument cannot be local, because this call is a tail call
```
(I wouldn't even bother mentioning the possibility of making the result `local_`: it's certainly a thing you can do, but it's unlikely to help someone who sees this error, so I wouldn't include it in the hint)

File: typing/typecore.ml [hunk 13]
```diff
@@ -5532,6 +5768,25 @@ let report_type_expected_explanation expl ppf =
...
+  | Types.Value_mode.Locality, Some Tailcall_function ->
+      [ Location.msg
+          "@[Hint: Local values cannot be called in tail position@ \
+           unless the result of the call has a \"local_\" annotation@]" ]
```
Comment (stedolan) [inline typing/typecore.ml]: ditto

**Review by stedolan — APPROVED** [review]
lgtm. New error messages are a big improvement.

File: typing/parmatch.ml (hunk 1 above)
Comment (lpw25) [reply typing/parmatch.ml]: I don't think it particularly matters, but maybe `max_mode` would be better. Not quite sure whether the dummy value should be the most restrictive or least restrictive.

File: typing/types.ml (hunk 5 above)
Comment (lpw25) [reply typing/types.ml:625]: The code already ensures that if `v.lower = v.upper` then `v.vlower` is the empty list. That means this fast path is not noticeably quicker than the slow path below.

File: typing/types.mli (hunk 10 above)
Comment (lpw25) [reply typing/types.mli]: I've had another go at renaming all these functions.

File: typing/env.ml (hunk 6 above)
Comment (lpw25) [reply typing/env.ml:2842]: Yes, it's because they are always global

File: typing/typecore.ml (hunk 9 above)
Comment (lpw25) [reply typing/typecore.ml:4765]: This only matters for known partial applications, which now don't complain about capturing the argument unless it is actually local.

**Review by stedolan — APPROVED** [review]
lgtm, good to merge once the conflict's fixed!

File: typing/typecore.ml (hunk 9 above)
Comment (stedolan) [reply typing/typecore.ml:4765]: Ah, I hadn't considered the partial case, that makes sense and is an improvement in behaviour.

I was slightly worried about the large number of new variables being created by this change. I checked, and this change increases the number of mode variables in a build of the compiler by about 40% ([results here](https://gist.github.com/stedolan/eb55c41680276d6390ed41e35387c722)). This doesn't seem to be much of a problem, but if we later run into problems with the mode variable solver this might be something to revisit.
